entity-framework/core/querying/null-semantics.md

```diff
@@ -0,0 +1,131 @@
+
+[!code-csharp[Main](../../../samples/core/Querying/NullSemantics/NullSemanticsContext.cs#UdfModelConfiguration)]
+
+First function is configured in a standard way. The second function is configured to take advantage of the nullability propagation optimization.
+
+When issuing the following queries:
+
+[!code-csharp[Main](../../../samples/core/Querying/NullSemantics/Program.cs#UdfExamples)]
+
+We get this SQL:
+
+```sql
+SELECT [e].[Id], [e].[Int], [e].[NullableInt], [e].[String1], [e].[String2]
+FROM [Entities] AS [e]
+WHERE [dbo].[ConcatStrings]([e].[String1], [e].[String2]) IS NOT NULL
+
+SELECT [e].[Id], [e].[Int], [e].[NullableInt], [e].[String1], [e].[String2]
+FROM [Entities] AS [e]
+WHERE [e].[String1] IS NOT NULL AND [e].[String2] IS NOT NULL
+```
+
+Just like with built-in `Substring` function, the second query doesn't need to evaluate the function itself to test its nullability.
+
+> [!NOTE]
+> This optimization should only be used if the function can only return `null` becuase it's parameters are `null`.
+
+## Optimizations and considerations
+
+- Comparing non-nullable columns is simpler and faster than comparing nullable columns. Consider marking columns as non-nullable wherever it is possible.
+
+- `Equal` comparison is simpler and faster than `NotEqual`, because query doesn't need to distinguish between `null` and `false` result. Use equality comparison whenever it's possible.
+
+> [!NOTE]
+> Wrapping `Equal` comparison around `Not` is effectively `NotEqual`. `Equal` comparison is only faster/simpler if it is not negated.
+
+- In some cases, it is possible to simplify a complex comparison by filtering out `null` values from a column explicitly - for example when no `null` values are present or these values are not relevant in the result. Consider the following example:
+
+[!code-csharp[Main](../../../samples/core/Querying/NullSemantics/Program.cs#ManualOptimization)]
+
+These queries produce the following SQL:
+
+```sql
+SELECT [e].[Id], [e].[Int], [e].[NullableInt], [e].[String1], [e].[String2]
+FROM [Entities] AS [e]
+WHERE ((([e].[String1] <> [e].[String1]) OR [e].[String1] IS NULL) AND [e].[String1] IS NOT NULL) OR ((CAST(LEN([e].[String1]) AS int) = CAST(LEN([e].[String1]) AS int)) OR [e].[String1] IS NULL)
+
+SELECT [e].[Id], [e].[Int], [e].[NullableInt], [e].[String1], [e].[String2]
+FROM [Entities] AS [e]
+WHERE [e].[String1] IS NOT NULL AND (CAST(LEN([e].[String1]) AS int) = CAST(LEN([e].[String1]) AS int))
+```
+
+In the second query, `null` results are filtered out from `String1` column explicitly. EF Core can safely treat the `String1` column as non-nullable during comparison, which results is simpler query.
+
+## Using relational null semantics
+
+It's possible to disable the null comparison compensation and use relational null semantics directly. This can be done by calling `UseRelationalNulls(true)` method on the options builder inside `OnConfiguring` method:
```

Comment:
We may want to add a warning such as:

> [!WARNING]
> When using relational null semantics, your LINQ queries no longer have the same meaning as they do in C#, and may yield different results than expected. Exercise caution when using this mode.

Reply:
I just responded to an issue about this and thought of the same thing. Some extra context/warning will be useful.